Sources/SyncDataProviders/Settings/SettingsProvider.swift

```diff
@@ -0,0 +1,294 @@
+        var saveAttemptsLeft = Const.maxContextSaveRetries
+
+        context.performAndWait {
+            while true {
+
+                do {
+                    let responseHandler = try SettingsResponseHandler(
+                        received: received,
+                        clientTimestamp: clientTimestamp,
+                        settingsHandlers: settingsHandlers,
+                        context: context,
+                        crypter: crypter,
+                        deduplicateEntities: isInitial
+                    )
+                    let idsOfItemsToClearModifiedAt = try cleanUpSentItems(sent, receivedKeys: Set(responseHandler.receivedByKey.keys), clientTimestamp: clientTimestamp, in: context)
+                    try responseHandler.processReceivedSettings()
+
+#if DEBUG
+                    try willSaveContextAfterApplyingSyncResponse()
+#endif
+                    let keys = idsOfItemsToClearModifiedAt.union(Set(responseHandler.receivedByKey.keys).subtracting(responseHandler.idsOfItemsThatRetainModifiedAt))
+                    try clearModifiedAtAndSaveContext(keys: keys, clientTimestamp: clientTimestamp, in: context)
+                    break
+                } catch {
+                    if (error as NSError).code == NSManagedObjectMergeError {
+                        context.reset()
+                        saveAttemptsLeft -= 1
+                        if saveAttemptsLeft == 0 {
+                            saveError = error
+                            break
+                        }
+                    } else {
+                        saveError = error
+                        break
+                    }
+                }
+
+            }
+        }
+        if let saveError {
+            throw saveError
+        }
+
+        if let serverTimestamp {
+            lastSyncTimestamp = serverTimestamp
+            syncDidUpdateData()
+        }
+    }
+
+    func cleanUpSentItems(_ sent: [Syncable], receivedKeys: Set<String>, clientTimestamp: Date, in context: NSManagedObjectContext) throws -> Set<String> {
+        if sent.isEmpty {
+            return []
+        }
+        let keys = sent.compactMap { SyncableSettingAdapter(syncable: $0).uuid }
+        let settingsMetadata = try SyncableSettingsMetadataUtils.fetchSettingsMetadata(for: keys, in: context)
+
+        var idsOfItemsToClearModifiedAt = Set<String>()
+
+        for metadata in settingsMetadata {
+            guard let adapter = settingsHandlers[Setting(key: metadata.key)] else {
+                continue
+            }
+
+            if let lastModified = metadata.lastModified, lastModified > clientTimestamp {
+                continue
+            }
+            let isLocalChangeRejectedBySync: Bool = receivedKeys.contains(metadata.key)
+            let isPendingDeletion = try adapter.getValue() == nil
+            if isPendingDeletion, !isLocalChangeRejectedBySync {
+                try adapter.setValue(nil)
```

Comment:
Setting nil value is good for Email Protection, but we will see if other settings follow this logic too. We can always abstract this to the handler protocol.